pkg/storage/gcp/saver.go

```diff
@@ -0,0 +1,92 @@
+package gcp
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/gobuffalo/envy"
+	"google.golang.org/api/option"
+)
+
+// Storage implements the (github.com/gomods/pkg/storage).Saver interface
+type Storage struct {
+	bucket *storage.BucketHandle
+}
+
+// New returns a new Storage instance authenticated using the provided
+// ClientOptions. The bucket name to be used will be loaded from the
+// environment variable ATHENS_GCP_BUCKET_NAME.
+//
+// The ClientOptions should provide permissions sufficient to read, write and
+// delete objects in google cloud storage for your project.
+func New(ctx context.Context, cred option.ClientOption) (*Storage, error) {
+	client, err := storage.NewClient(ctx, cred)
+	if err != nil {
+		return nil, fmt.Errorf("could not create new client: %s", err)
+	}
+	bucketname, err := envy.MustGet("ATHENS_GCP_BUCKET_NAME")
+	if err != nil {
+		return nil, fmt.Errorf("could not load 'ATHENS_GCP_BUCKET_NAME': %s", err)
+	}
+	bkt := client.Bucket(bucketname)
+	return &Storage{bucket: bkt}, nil
+}
+
+// Save uploads the modules .mod, .zip and .info files for a given version
+func (s *Storage) Save(ctx context.Context, module, version string, mod, zip, info []byte) error {
+	errs := make(chan error, 3)
+	// create a context that will time out after 300 seconds / 5 minutes
+	ctxWT, _ := context.WithTimeout(ctx, 300*time.Second)
```

Comment:
please use defer cancel(), it removes itself from parent context (clears resources)

Reply:
awesome, thanks. I was wondering about that.